api/conversion/conversion.go

```diff
@@ -0,0 +1,191 @@
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	// AdminUserID represents the default user id of service.
+	AdminUserID string = "admin"
+
+	// AdminUsername represents the default username of service.
+	AdminUsername string = "admin"
+)
+
+// ConvertPipelineToService converts the pipeline to service, as the running of pipeline dependents on service. This
+// is just a workaround, the pipeline will can be run directly in the future.
+func ConvertPipelineToService(pipeline *newapi.Pipeline) (*api.Service, error) {
+	service := &api.Service{}
+
+	// Basic information of service.
+	service.Name = pipeline.Name
+	service.Description = pipeline.Description
+
+	if pipeline.Build == nil && pipeline.Build.Stages == nil && pipeline.Build.Stages.CodeCheckout == nil {
+		return nil, fmt.Errorf("fail to generate service as code checkout stages is empty")
+	}
+
+	// Convert the repository for the service.
+	repository, err := convertRepository(pipeline.Build.Stages.CodeCheckout)
+	if err != nil {
+		return nil, err
+	}
+	service.Repository = *repository
+
+	service.UserID = AdminUserID
+	// Username is used as the repo of built image.
+	service.Username = AdminUsername
+
+	// Convert the build stages to caicloud YAML string for service.
+	caicloudYamlStr, err := convertBuildStagesToCaicloudYaml(pipeline)
+	if err != nil {
+		return nil, err
+	}
+	service.CaicloudYaml = caicloudYamlStr
+
+	return service, nil
+}
+
+// convertBuildStagesToCaicloudYaml converts the config of build stages in pipeline to caicloud YAML.
+func convertBuildStagesToCaicloudYaml(pipeline *newapi.Pipeline) (string, error) {
+	if pipeline.Build == nil || pipeline.Build.BuilderImage == nil || pipeline.Build.Stages == nil {
+		return "", fmt.Errorf("fail to generate caicloud YAML as builder image or build stages is empty")
+	}
+
+	builderImage := pipeline.Build.BuilderImage
+	stages := pipeline.Build.Stages
+	if stages.Package == nil {
+		return "", fmt.Errorf("fail to generate caicloud YAML as package stages is empty")
+	}
+
+	if stages.ImageBuild == nil {
+		return "", fmt.Errorf("fail to generate caicloud YAML as image build stages is empty")
+	}
+
+	caicloudYAMLConfig := &Config{}
+
+	// Convert the package stage of pipeline to the prebuild of caicloud YAML.
+	preBuild := &PreBuild{}
+	packageConfig := stages.Package
+	preBuild.Commands = packageConfig.Command
+	preBuild.Outputs = packageConfig.Outputs
+	preBuild.Image = builderImage.Image
+	preBuild.Environment = convertEnvVars(builderImage.EnvVars)
+
+	caicloudYAMLConfig.PreBuild = preBuild
+
+	// Convert the image build stage to the build of caicloud YAML.
+	build := &Build{}
+	imageBuildConfig := stages.ImageBuild
+	buildInfoNum := len(imageBuildConfig.BuildInfos)
+	if buildInfoNum == 0 || buildInfoNum > 1 {
+		return "", fmt.Errorf("fail to generate caicloud YAML as %d buildInfos provided in imageBuild stage", buildInfoNum)
+	}
+
+	// Now only support one build info.
+	// TODO (robin) How to support the customized Dockerfile.
```

Comment:
@zoumo had achieved this feature in Q1. There is a element called `Dockerfile` in service. If `Dockerfile` is empty, cyclone will use the Dockerfile in repo, else cyclone will use this element.

Reply:
There are only fields `DockerfilePath` and `DockerfileName`, no `Dockerfile`.

Reply:
https://github.com/caicloud/cyclone/blob/master/api/types.go#L86